From 92c1d53f95d9be00d94097a578dc963d51e837f0 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:42:35 -0400
Subject: [ethernet] cxgb4: Fix some small bugs in t4_sge_init_soft() when our Page Size is 64KB

Message-id: <1410475447-94392-17-git-send-email-svemuri@redhat.com>
Patchwork-id: 93807
O-Subject: [RHEL7.1 PATCH BZ 1124947 016/108] cxgb4: Fix some small bugs in t4_sge_init_soft() when our Page Size is 64KB
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

We'd come in with SGE_FL_BUFFER_SIZE[0] and [1] both equal to 64KB and the
extant logic would flag that as an error.

Based on original work by Casey Leedom <leedom@chelsio.com>

Signed-off-by: Kumar Sanghvi <kumaras@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 92ddcc7b8f1c14aa9f3ec98b14bcd421b21b01e4)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/sge.c b/drivers/net/ethernet/chelsio/cxgb4/sge.c
index e12fba6..332212c 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/sge.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/sge.c
@@ -2594,11 +2594,19 @@ static int t4_sge_init_soft(struct adapter *adap)
  fl_small_mtu = READ_FL_BUF(RX_SMALL_MTU_BUF);
  fl_large_mtu = READ_FL_BUF(RX_LARGE_MTU_BUF);
 
+ /* We only bother using the Large Page logic if the Large Page Buffer
+  * is larger than our Page Size Buffer.
+  */
+ if (fl_large_pg <= fl_small_pg)
+  fl_large_pg = 0;
+
  #undef READ_FL_BUF
 
+ /* The Page Size Buffer must be exactly equal to our Page Size and the
+  * Large Page Size Buffer should be 0 (per above) or a power of 2.
+  */
  if (fl_small_pg != PAGE_SIZE ||
-     (fl_large_pg != 0 && (fl_large_pg < fl_small_pg ||
-      (fl_large_pg & (fl_large_pg-1)) != 0))) {
+     (fl_large_pg & (fl_large_pg-1)) != 0) {
   dev_err(adap->pdev_dev, "bad SGE FL page buffer sizes [%d, %d]\n",
    fl_small_pg, fl_large_pg);
   return -EINVAL;
-- 
1.7.1